modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessExecuteAction.java

```diff
@@ -86,31 +111,76 @@ public Response newResponse() {
-                assert token == XContentParser.Token.END_OBJECT;
-                token = p.nextToken();
-                assert token == XContentParser.Token.END_OBJECT;
-                return SupportedContext.valueOf(contextType.toUpperCase(Locale.ROOT));
+                SupportedContext supportedContext = SupportedContext.valueOf(contextType.toUpperCase(Locale.ROOT));
+                if (supportedContext == SupportedContext.PAINLESS_TEST) {
+                    token = p.nextToken();
+                    assert token == XContentParser.Token.START_OBJECT;
+                    token = p.nextToken();
+                    assert token == XContentParser.Token.END_OBJECT;
+                    token = p.nextToken();
+                    assert token == XContentParser.Token.END_OBJECT;
+                } else if (supportedContext == SupportedContext.FILTER_SCRIPT || supportedContext == SupportedContext.SEARCH_SCRIPT) {
+                    for (token = p.nextToken(); token != XContentParser.Token.END_OBJECT; token = p.nextToken()) {
+                        if (token.isValue()) {
+                            if (INDEX_FIELD.match(p.currentName(), p.getDeprecationHandler())) {
+                                c.index = p.textOrNull();
+                            } else {
+                                throw new IllegalStateException("Unexpected field name [" + p.currentName() + "]");
+                            }
+                        } else if (token == XContentParser.Token.START_OBJECT) {
+                            if (DOCUMENT_FIELD.match(p.currentName(), p.getDeprecationHandler())) {
+                                try (XContentBuilder builder = XContentFactory.jsonBuilder()) {
+                                    builder.copyCurrentStructure(p);
+                                    builder.flush();
+                                    c.document = BytesReference.bytes(builder);
+                                }
+                            }
+                        } else if (token == XContentParser.Token.FIELD_NAME) {
+                            // ignore
+                        } else {
+                            throw new IllegalStateException("Unexpected token [" + token + "]");
+                        }
+                    }
+                } else {
+                    assert false : "Unsupported context";
+                }
+                return supportedContext;
             }, CONTEXT_FIELD);
         }
 
         private Script script;
         private SupportedContext context;
 
+        private String index;
+        private BytesReference document;
+        private XContentType xContentType;
+
         static Request parse(XContentParser parser) throws IOException {
-            return PARSER.parse(parser, null);
+            ParseContext parseContext = new ParseContext();
+            Request request = PARSER.parse(parser, parseContext);
+            request.setIndex(parseContext.index);
```

Comment:
Isn't the idea of using ObjectParser so the object returned from the parser is completely constructed?

Reply:
That is true, but that couldn't be done in this case. In the json the document / index name are provided under the context field, which maps the `context` field in the request class, but that is an enum. So via this way we attach the index name and document to the parse context, so that we later apply them to the `document` and `index` fields in the request class.